### PR TITLE
Add locale parameter to EditableText

### PR DESCRIPTION
**Pull request by HansMuller**
The RenderParagraph and RenderEditableText renderers now pass the locale along to the engine via their internal TextPainters (see also https://github.com/flutter/flutter/pull/18189, https://github.com/flutter/flutter/pull/17879). The locale is typically defined by the app and text widgets look it up with Localizations.localeOf().

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -249,6 +251,22 @@ class RenderEditable extends RenderBox {
...
 
+  /// Used by this renderer's internal [TextPainter] to select a locale-specific
+  /// font.
+  ///
+  /// In some cases the same Unicode character may be rendered differently depending
+  /// on the locale. For example the '骨' character is rendered differently in
+  /// the Chinese and Japanese locales. In these cases the [locale] may be used
+  /// to select a locale-specific font.
+  Locale get locale => _textPainter.locale;
+  /// The value may be null.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: we should probably mention briefly what null means

**Comment by Hixie** [comment]
As discussed offline, I realized while reviewing this patch that we do need to have the Locale on the TextStyle, because a single Text widget can have multiple parts of its text with different locales, e.g. "The English word foo is bar in Japanese and baz in Chinese" would have three locales.

The TextStyle should probably default to null locale and then Text can fill it in from the ambient Locale.

**Comment by HansMuller** [comment]
> The TextStyle should probably default to null locale and then Text can fill it in from the ambient Locale.

If you create a RichText with TextSpans whose TextStyles have non-null locales, the RichText will carry the ambient locale and the TextSpan styles will carry along their locales via the ui.ParagraphBuilder.

In this case the engine uses the text style's locale.

I've verified that this works. Will add tests when tests can use fonts.
```
    final String character = '骨';
    final TextStyle style = Theme.of(context).textTheme.display3;
    return new Scaffold(
      body: new Container(
        padding: const EdgeInsets.all(48.0),
        alignment: Alignment.center,
        child: new RichText(
          text: new TextSpan(
            children: <TextSpan>[
              new TextSpan(text: character, style: style.copyWith(locale: const Locale('ja'))),
              new TextSpan(text: character, style: style.copyWith(locale: const Locale('zh'))),
            ],
          ),
        ),
      ),
    );
```

**Comment by Hixie** [comment]
Can you add the test anyway today (as a golden test) and then add a TODO to enable it, with a link to the relevant font bug? That way when we implement it we'll have a better chance to also update the test.

File: packages/flutter/lib/src/painting/text_style.dart [hunk 2]
```diff
@@ -299,6 +299,13 @@ class TextStyle extends Diagnosticable {
   final double height;
 
   /// The locale used to select region-specific glyphs.
+  ///
+  /// This property is rarely set. Typically the locale used to select
+  /// region-specific glyphs is defined by the text widget's [BuildContext]
+  /// using `Localizations.localeOf(context)`. For example [RichText] defines
+  /// its locale this way. However a rich text widget's [TextSpan]s could specify
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/text_style.dart]: nit: comma after however

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -249,6 +251,24 @@ class RenderEditable extends RenderBox {
...
 
+  /// Used by this renderer's internal [TextPainter] to select a locale-specific
+  /// font.
+  ///
+  /// In some cases the same Unicode character may be rendered differently depending
+  /// on the locale. For example the '骨' character is rendered differently in
+  /// the Chinese and Japanese locales. In these cases the [locale] may be used
+  /// to select a locale-specific font.
+  ///
+  /// If this value is null, a system dependent algorithm is used to select
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/editable.dart]: nit: system-dependent

**Comment by Hixie** [comment]
LGTM

We should also add a test that checks that when the text style has a locale and the text has a locale, the text style one isn't replaced.

**Comment by HansMuller** [comment]
I've added golden image tests that will run and succeed with the Ahem font.  When https://github.com/flutter/flutter/issues/17700 is fixed, the golden files will need to be updated

**Pull request merged**